python/cudf/cudf/core/dataframe.py

```diff
@@ -401,57 +405,55 @@ class _DataFrameIlocIndexer(_DataFrameIndexer):
     For selection by index.
     """
 
-    @_cudf_nvtx_annotate
-    def _getitem_tuple_arg(self, arg):
-        # Iloc Step 1:
-        # Gather the columns specified by the second tuple arg
-        columns_df = self._frame._from_data(
-            self._frame._data.select_by_index(arg[1]), self._frame._index
-        )
+    _frame: DataFrame
 
-        # Iloc Step 2:
-        # Gather the rows specified by the first tuple arg
-        if isinstance(columns_df.index, MultiIndex):
-            if isinstance(arg[0], slice):
-                df = columns_df[arg[0]]
-            else:
-                df = columns_df.index._get_row_major(columns_df, arg[0])
-            if (len(df) == 1 and len(columns_df) >= 1) and not (
-                isinstance(arg[0], slice) or isinstance(arg[1], slice)
-            ):
-                # Pandas returns a numpy scalar in this case
-                return df.iloc[0]
-            if self._can_downcast_to_series(df, arg):
-                return self._downcast_to_series(df, arg)
-            return df
+    def __getitem__(self, arg):
+        row_key, (
+            col_is_scalar,
+            column_names,
+        ) = iu.destructure_dataframe_iloc_indexer(arg, self._frame)
+        row_spec = iu.parse_row_iloc_indexer(
+            row_key, len(self._frame), check_bounds=True
+        )
+        ca = self._frame._data
+        index = self._frame.index
+        if col_is_scalar:
+            s = Series._from_data(
+                ca._select_by_names(column_names), index=index
+            )
+            return s._getitem_preprocessed(row_spec)
+        if column_names != list(self._frame._column_names):
+            frame = self._frame._from_data(
+                ca._select_by_names(column_names), index=index
+            )
         else:
-            if isinstance(arg[0], slice):
-                df = columns_df._slice(arg[0])
-            elif is_scalar(arg[0]):
-                index = arg[0]
-                if index < 0:
-                    index += len(columns_df)
-                df = columns_df._slice(slice(index, index + 1, 1))
-            else:
-                arg = (as_column(arg[0]), arg[1])
-                if is_bool_dtype(arg[0]):
-                    df = columns_df._apply_boolean_mask(arg[0])
-                else:
-                    df = columns_df._gather(arg[0])
-
-        # Iloc Step 3:
-        # Reindex
-        if df.shape[0] == 1:  # we have a single row without an index
-            df.index = as_index(self._frame.index[arg[0]])
-
-        # Iloc Step 4:
-        # Downcast
-        if self._can_downcast_to_series(df, arg):
-            return self._downcast_to_series(df, arg)
-
-        if df.shape[0] == 0 and df.shape[1] == 0 and isinstance(arg[0], slice):
-            df._index = as_index(self._frame.index[arg[0]])
-        return df
+            frame = self._frame
+        if isinstance(row_spec, iu.MapIndexer):
```

Comment:
With the typing that you've added it could be helpful to extract this logic into a `functools.singledispatch` helper function.

Reply:
I thought about that, but it has the slight downside that the typecheckers can't see through the dispatch registration to determine that the base registration's `assert_never(spec)` is indeed unreachable:

https://mypy-play.net/?mypy=latest&python=3.11&gist=1a7dcd4f626de70aae853467666dc4b2

You might notice that this dispatch is _almost_ identical to `Series._getitem_preprocessed` apart from the attempt to downcast to a 1xN result to a series.... Should really propose a deprecation in pandas.